sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/TransactionalBatch.java

```diff
@@ -0,0 +1,306 @@
+ *     .createItem<ToDoActivity>(test1)
+ *     .replaceItem<ToDoActivity>(test2.id, test2)
+ *     .upsertItem<ToDoActivity>(test3)
+ *     .deleteItem("reading");
+ *
+ * try (TransactionalBatchResponse response = container.executeTransactionalBatch(batch) {
+ *
+ *     if (!response.IsSuccessStatusCode) {
+ *        // Handle and log exception
+ *        return;
+ *     }
+ *
+ *     // Look up interested results - e.g., via typed access on operation results
+ *
+ *     TransactionalBatchOperationResult<ToDoActivity> result = response.getOperationResultAtIndex<ToDoActivity>(0, ToDoActivity.class);
+ *     ToDoActivity readActivity = result.getItem();
+ * }
+ * }</pre>
+ *
+ * <b>Example</b>
+ * <p>This example atomically reads a set of documents as a batch.
+ * <pre>{@code
+ * String activityType = "personal";
+ *
+ * TransactionalBatch batch = TransactionalBatch.createTransactionalBatch(new Cosmos.PartitionKey(activityType))
+ *     .readItem("playing")
+ *     .readItem("walking")
+ *     .readItem("jogging")
+ *     .readItem("running")
+ *
+ * try (TransactionalBatchResponse response = container.executeTransactionalBatch(new Cosmos.PartitionKey(activityType)) {
+ *
+ *     // Look up interested results - eg. via direct access to operation result stream
+ *
+ *     List<String> resultItems = new ArrayList<String>();
+ *
+ *     for (TransactionalBatchOperationResult result : response) {
+ *         resultItems.add(result.getResourceObject().toString())
+ *     }
+ * }
+ * }</pre>
+ * <p>
+ * <b>See:</b>
+ * <a href="https://docs.microsoft.com/azure/cosmos-db/concepts-limits">Limits on TransactionalBatch requests</a>.
+ */
+public final class TransactionalBatch {
+
+    private final ArrayList<ItemBatchOperation<?>> operations;
+    private final PartitionKey partitionKey;
+
+    public TransactionalBatch(PartitionKey partitionKey) {
+        this.operations = new ArrayList<>();
+        this.partitionKey = partitionKey;
+    }
+
+    /**
+     * Initializes a new instance of {@link TransactionalBatch}
+     * that will contain operations to be performed across multiple items in the container with the provided partition
+     * key in a transactional manner
+     *
+     * @param partitionKey the partition key for all items in the batch.
+     *
+     * @return A new instance of {@link TransactionalBatch}.
+     */
+    public static TransactionalBatch createTransactionalBatch(PartitionKey partitionKey) {
+        return new TransactionalBatch(partitionKey);
+    }
+
+    /**
+     * Adds an operation to create an item into the batch.
+     *
+     * @param item A JSON serializable object that must contain an id property.
+     * @param <T> The type of item to be created.
+     *
+     * @return The transactional batch instance with the operation added.
+     */
+    public <T> TransactionalBatch createItem(T item) {
+        checkNotNull(item, "expected non-null item");
```

Comment:
I think usually we throw an IllegalArguemntException when doing null-validation of parameters in public surface area - this one will throw a NullPointerException instead. 

@moderakh / @kushagraThapar - I haven't found clear guidance on this in the Java world in the past (NPE vs. IllegalArgumentException for null-checks). Are the Azure SDK common guidelines? If not do you have a strong preference for one over the other?

Reply:
Discussed in-person - we aren't consistent right now - tendency was to use NullPointerException instead of IllegalArgumentException going forward.